security/advancedtls/provider.go

```diff
@@ -0,0 +1,201 @@
+	if o.Interval == 0 {
+		o.Interval = defaultInterval
+	}
+	ticker := time.NewTicker(o.Interval)
+	ctx, cancel := context.WithCancel(context.Background())
+	// Initialize the distributor of provider with an empty KeyMaterial.
+	provider := &IdentityPEMFileProvider{distributor: certprovider.NewDistributor()}
+	provider.distributor.Set(&certprovider.KeyMaterial{}, nil)
+	// A goroutine to pull file changes.
+	go func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				{
+					// Read identity certs from PEM files.
+					identityCert, err := tls.LoadX509KeyPair(o.CertFile, o.KeyFile)
+					if err != nil {
+						// If the reading produces an error, we will skip the update for this round and log the error.
+						// Note that LoadX509KeyPair will return error if file is empty,
+						// so there is no separate check for empty file contents.
+						logger.Warning("tls.LoadX509KeyPair(%v, %v) failed: %v", o.CertFile, o.KeyFile, err)
+						continue
+					}
+					provider.distributor.Set(&certprovider.KeyMaterial{Certs: []tls.Certificate{identityCert}}, nil)
+				}
+			default:
+			}
+		}
+	}(ctx)
+	provider.cancel = cancel
+	return provider, nil
+}
+
+// KeyMaterial returns the key material sourced by the IdentityPEMFileProvider.
+// Callers are expected to use the returned value as read-only.
+func (p *IdentityPEMFileProvider) KeyMaterial(ctx context.Context) (*certprovider.KeyMaterial, error) {
+	return p.distributor.KeyMaterial(ctx)
+}
+
+// Close cleans up resources allocated by the IdentityPEMFileProvider.
+func (p *IdentityPEMFileProvider) Close() {
+	p.cancel()
+	p.distributor.Stop()
+}
+
+// NewRootPEMFileProvider uses RootPEMFileProviderOptions to construct a RootPEMFileProvider.
+func NewRootPEMFileProvider(o *RootPEMFileProviderOptions) (*RootPEMFileProvider, error) {
+	if o.TrustFile == "" {
+		return nil, fmt.Errorf("users must specify TrustFile in RootPEMFileProviderOptions")
+	}
+	// If interval is not set by users explicitly, we will set it to default interval.
+	if o.Interval == 0 {
+		o.Interval = defaultInterval
+	}
+	ticker := time.NewTicker(o.Interval)
+	ctx, cancel := context.WithCancel(context.Background())
+	// Initialize the distributor of provider with an empty KeyMaterial.
+	provider := &RootPEMFileProvider{distributor: certprovider.NewDistributor()}
+	provider.distributor.Set(&certprovider.KeyMaterial{}, nil)
+	// A goroutine to pull file changes.
+	go func(ctx context.Context) {
+		for {
+			select {
+			case <-ctx.Done():
+				ticker.Stop()
+				return
+			case <-ticker.C:
+				{
+					if trustFileSize, _ := getFileSize(o.TrustFile); trustFileSize == 0 {
+						// If the current file is empty, skip the update for this round.
+						continue
+					}
+					// Read root certs from PEM files.
+					trustData, err := ioutil.ReadFile(o.TrustFile)
+					trustPool := x509.NewCertPool()
+					trustPool.AppendCertsFromPEM(trustData)
```

Comment:
Let's check the result of `AppendCertsFromPEM `

Reply:
Sure!

Reply:
Is it okay to check `trustData` since that's the result from reading files? In this way, we will avoid executing the next two lines if it is empty.